src/TextInputWidget.cc

```diff
@@ -135,28 +135,42 @@ FilteredTextEdit::canInsertFromMimeData(const QMimeData *source) const
 void
 FilteredTextEdit::insertFromMimeData(const QMimeData *source)
 {
-        if (source->hasImage()) {
-                const auto formats = source->formats();
-                const auto idx     = formats.indexOf(
-                  QRegularExpression{"image/.+", QRegularExpression::CaseInsensitiveOption});
-
-                // Note: in the future we may want to look into what the best choice is from the
-                // formats list. For now we will default to PNG format.
-                QString type = "png";
-                if (idx != -1) {
-                        type = formats.at(idx).split('/')[1];
+        const auto formats = source->formats().filter("/");
+        const auto image   = formats.filter("image/", Qt::CaseInsensitive);
+        const auto audio   = formats.filter("audio/", Qt::CaseInsensitive);
+        const auto video   = formats.filter("video/", Qt::CaseInsensitive);
+
+        if (!image.empty()) {
+                showPreview(source, image);
+        } else if (!audio.empty()) {
+                showPreview(source, audio);
+        } else if (!video.empty()) {
+                showPreview(source, video);
+        } else if (source->hasUrls()) {
+                QString path;
+                if (QFileInfo{source->text()}.exists()) {
+                        path = source->text();
+                } else {
+                        for (auto &&u : source->urls()) {
+                                if (u.isLocalFile()) {
+                                        path = u.toLocalFile();
+                                        break;
+                                }
+                        }
                 }
 
-                // Encode raw pixel data of image.
-                QByteArray data = source->data("image/" + type);
-                previewDialog_.setImageAndCreate(data, type);
-                previewDialog_.show();
-        } else if (source->hasFormat("x-special/gnome-copied-files") &&
-                   QImageReader{source->text()}.canRead()) {
+                if (!path.isEmpty() && QFileInfo{path}.exists()) {
+                        previewDialog_.setPreview(path);
+                } else {
+                        qWarning()
+                          << "Clipboard does not contain any valid file paths:" << source->urls();
+                }
+        } else if ((source->hasText() || source->hasFormat("x-special/gnome-copied-files")) &&
+                   QFileInfo{source->text()}.exists()) {
```

Comment:
What prevents this from triggering if I paste plain text that happens to be a path to a local file?

Reply:
Ah good catch. That was an oversight.